Handle rented node in grace period

samaradel opened the pull request:
### Description

Check if the rented node contract is in the grace period when listing the node.

### Changes

![image](https://github.com/user-attachments/assets/1f1b3866-ee34-4737-84bf-b43fbef80cbe)

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3025

### Tested Scenarios

- Navigate to the Node finder page.
- Rent a dedicated node.
- Leave it till it goes to the grace period.
- Navigate to any solution.
- Try to deploy an instance on the rented node.

### Checklist

- [x] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

Commit:
Handle rented node in grace period

Commit:
Handle Error

Commit:
Remove the error toast

Commit:
Remove unused imports

Commit:
Add timeout with promise to catch grace period error

Commit:
Remove umwanted imports

samaradel commented:
cpd from https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3646

Commit:
Handle console error

0oM4R reviewed (commented):
![image](https://github.com/user-attachments/assets/87581bb5-5060-4d81-a904-f30e9dba31fa)


https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3646#pullrequestreview-2504471530

samaradel commented:
@0oM4R check it, it already fixed

amiraabouhadid reviewed (approved):
no console err 
![image](https://github.com/user-attachments/assets/30ba6a57-8ca9-4b79-b126-37762a303438)

Commit:
Merge branch 'development' into development_2.7_handle_rentedNode

0oM4R reviewed (commented):
on first if the node is selected; 
we won't see any error but if you select another node then click on the graceperiod node will see the error


also the error must be handled 

- the node is not a valid node; so we shouldn't auto select it; need to handle this case as well 




![image](https://github.com/user-attachments/assets/d5483656-9649-4842-be1e-2281755d8326)

0oM4R commented on packages/playground/src/utils/nodeSelector.ts:
```diff
@@ -235,10 +234,13 @@ export async function validateRentContract(
-        );
+        const err = `You can't deploy on node ${node.nodeId}, its rent contract is in grace period.`;
+        await new Promise((_, reject) => {
+          setTimeout(() => {
+            reject(Error(err));
+          }, 2000);
+        });
```
i don't think we need to add timeout here, not the auto one there is no toast overlap what do you think?

samaradel replied:
totally agree, removed

0oM4R replied:
I see you removed timeout in the automatic on to, can you explain why I think ti is a better UX even if we set it to only one sec

Commit:
Merge branch 'development' into development_2.7_handle_rentedNode

Commit:
rm timeout

0oM4R reviewed (changes requested):
![graceperiod](https://github.com/user-attachments/assets/fb377ad3-ebd9-494e-a113-e264aa3618ea)


error still exsist if i select the grace period node

samaradel commented:
@0oM4R works with me, please make sure to have the latest update and build

0oM4R commented:
https://github.com/user-attachments/assets/f8faaada-8999-4d4d-8585-6b940c0b4958

Commit:
Merge branch 'development' into development_2.7_handle_rentedNode

Commit:
Fix recursive calls

0oM4R commented on packages/playground/src/components/node_selector/TfAutoNodeSelector.vue:
```diff
@@ -304,14 +303,24 @@ export default {
         }
+
         const nodeCapacityValid = await checkNodeCapacityPool(gridStore, node, props.filters);
         const rentContractValid = props.filters.dedicated ? await validateRentContract(gridStore, node) : true;
```
we already do the same in validateRentContract is there any extra logic added in the new added lines above

samaradel replied:
I see it only fires when `props.filters.dedicated` is on and with our latest changes in the dashboard we can have a  dedicated node in the normal list, not sure if we should remove it

0oM4R replied:
so we should change this condition to be applicable with the latest changes; My point is we shouldn't do any validation inside the details card; in some cases we will do the same validation twice! on inside the details card and the other in the parent component

samaradel replied:
agreed

0oM4R commented on packages/playground/src/components/node_selector/TfNodeDetailsCard.vue:
```diff
@@ -364,6 +356,18 @@ export default {
       tftsNeeded();
     });
 
+    async function handleNodeClick() {
+      try {
+        if (props.status === "Init" && props.node) {
+          ctx.emit("node:select", props.node as NodeInfo);
+        }
+        if (props.node?.dedicated && props.node.rentContractId === 0) return false;
+        await validateRentContract(gridStore, props.node as NodeInfo);
+      } catch (error) {
+        return error;
+      }
+    }
+
```
why we are calling `validateRentContract` here while it is already on the manual and auto node finder?
![image](https://github.com/user-attachments/assets/b8064b4f-bc43-4011-be27-a0c6b309a827)
i think we need to make it just emit the select event and the parent component will handle the validation

samaradel replied:
same reason https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3742/files#r1966744835

0oM4R commented on packages/playground/src/utils/nodeSelector.ts:
```diff
@@ -231,14 +230,11 @@ export async function validateRentContract(
       throw `Node ${node.nodeId} is not rented`;
     }
     if (node.rentContractId !== 0) {
```
I suggest having early return here if the node is not rented and not in a dedicated farm, so there is no need to continue, right?

just a reminder that the dedicated node in shared farm can be used as shared node

0oM4R reviewed (changes requested):
I think we just need the changes in utils only; no need for the other changes

Commit:
q
- Remove unwanted contract validation
- Early exit for check it node is rented and not in dedicated farm
- Remove the check only if dedicated filter is ON

0oM4R reviewed (approved):
dedicated nodes in shared farms are valid nodes 
<img width="1263" alt="image" src="https://github.com/user-attachments/assets/75710324-9621-43d6-aab1-9d6072e7b5e2" />

validation works fine on automatic finder 

<img width="1302" alt="image" src="https://github.com/user-attachments/assets/4f8e6399-2b17-4e00-bac1-cad2beebd75e" />


all works fine thanks.


<img width="1249" alt="image" src="https://github.com/user-attachments/assets/69784fbb-9684-4525-94e7-a4299e1cc6ec" />
